### PR TITLE
Add buffered output support to console subsystem.

### PR DESCRIPTION
**Pull request by pfalcon**
This implements completely interrupt-driven UART console in Zephyr.

The result of a number of patches to fix/cleanup/better define semantics of UART API in Zephyr.

### CODE REVIEW

**Comment by pfalcon** [comment]
This is "top-level goal" ticket for a number of other changes, e.g. #42, #43.

In the present shape, it works only on STM32, and need further work as described in https://jira.zephyrproject.org/browse/ZEP-2016

File: subsys/console/getchar.c [hunk 1]
```diff
@@ -14,22 +14,94 @@
...
+//	uart_irq_update(dev);
+
+	if (uart_irq_rx_ready(dev)) {
+		char c;
+
+		uart_fifo_read(dev, &c, 1);
+		console_irq_input_hook(c);
+		//uart_irq_rx_disable(dev);
+	}
+	if (uart_irq_tx_empty(dev)) {
```
Comment (tbursztyka) [inline subsys/console/getchar.c]: break a line

Comment (pfalcon) [inline subsys/console/getchar.c]: Done

**Comment by pfalcon** [comment]
(Un?)surprisingly this also works on frdm_k64f, which confirms there're wider problems than just stm32.

**Comment by pfalcon** [comment]
RFC on clarifying tx_ready vs tx_empty was posted, and in its spirit, this patch was switched to tx_ready. Also, assuming that rx_ready *can* be edge interrupt, so its handler should try to read as much input as possible with uart_fifo_read().

In this shape, "echo" sample (included) was tested on: 96b_carbon, frdm_k64f, arduino_101, and behaves as expected.

**Comment by pfalcon** [comment]
RFC on clarifying tx_ready vs tx_empty was posted: https://lists.zephyrproject.org/pipermail/zephyr-devel/2017-May/007642.html, and in its spirit, this patch was switched to tx_ready. Also, assuming that rx_ready *can* be edge interrupt, so its handler should try to read as much input as possible with uart_fifo_read().

In this shape, "echo" sample (included) was tested on: 96b_carbon, frdm_k64f, arduino_101, and behaves as expected.

The debug output is still in this patch. Next step is to get feedback on RFC above, submit patch implementing it, and then finally this patch can be cleaned up.

**Comment by pfalcon** [comment]
With https://github.com/zephyrproject-rtos/zephyr/pull/159 being ready for merge, this patch has been cleaned up, and now should be ready for the final review too.

@daniel-thompson , @galak , @tbursztyka , @erwango : Please review.

**Comment by pfalcon** [comment]
@nashif : So, this mostly adds new functionality (with minor refactors) to the console subsystem, which was created by me in the first place. As a subsys maintainer I thus approve it ;-). If code looks ok for you, feel free to merge it.

**Review by erwango — APPROVED** [review]
LGTM and tested ok on disco_l475_iot1 (stm32l4)

File: include/console.h [hunk 2]
```diff
@@ -13,16 +13,17 @@
...
- *  supersedes, and incompatible with, callback (push-style) console
- *  handling (via console_input_fn callback, etc.).
+ *  access to console via console_getchar() function and buffered
+ *  output using console_putchar() function. This function supercedes,
+ *  and incompatible with, callback (push-style) console handling
+ *  (via console_input_fn callback, etc.).
  *
  *  @return N/A
  */
-void console_getchar_init(void);
```
Comment (galak) [inline include/console.h]: Is there a concern about us removing an API, not sure if we can support a deprecated version or not?

Comment (pfalcon) [inline include/console.h]: It was added in very this development cycle, so it was never released. As I mentioned in another comment, the whole "interrupt-driven console" subsystem is authored by me.

**Review by nashif — APPROVED** [review]

**Pull request merged**